### PR TITLE
[message] add range-based `for` loop for message/priority queue

### PR DESCRIPTION
**Pull request by abtink**
This commit adds support for range-based `for` loop iteration over
`MessageQueue` and `PriorityQueue`. It adds `ConstIterator` and
`Iterator` in `Message` class. The non-const `for` loop iteration
(which uses `Message::Iterator`) works properly and is safe to use
even when the entry is removed from the queue during iteration. This
commit updates other core modules to use the new `for` loop iteration
model which helps simplify the code. It also updates the unit tests
to validate the behavior of the newly added iteration mechanisms.

----

The non-const version implements and hides the common pattern where we 
update and remember the `nextMessage` at the start of loop iteration (so 
we can remove the `message` entry from the queue while iterating over it), into 
the `Message::Iterator` class itself.

So the common pattern
```c++
Message *message, 
Message *nextMessage;

for (message = aQueue.GetHead(); message != nullptr; message = nextMessage)
{
    nextMessage = message->GetNext();
    ...
}
```
can be replaced with:
```c++
for (Message &message: aQueue)
{
    ...
}
```

### CODE REVIEW

File: src/core/common/message.hpp [hunk 1]
```diff
@@ -1225,6 +1227,45 @@ class Message : public otMessage, public Buffer
...
+    public:
+        ConstIterator(void) = default;
+
+        explicit ConstIterator(const Message *aMessage)
+            : ItemPtrIterator(aMessage)
+        {
+        }
+
+    private:
+        void Advance(void) { mItem = (mItem != nullptr) ? mItem->GetNext() : nullptr; }
```
Comment (simonlingoogle) [inline src/core/common/message.hpp:1243]:
```suggestion
        void Advance(void) { mItem = mItem->GetNext(); }
```

Should `Advance()` never be called when mItem is nullptr ?

Comment (abtink) [reply src/core/common/message.hpp:1243]: Yes. I think we can remove this. 
If `Iterator` is at end, calling `++` or `Advance()` on it is usually undefined behavior (incorrect use).

Comment (abtink) [reply src/core/common/message.hpp]: Changed this in new push.

**Review by simonlingoogle — APPROVED** [review]
LGTM 👍 

thanks

**Review by jwhui — APPROVED** [review]

**Pull request merged**